File: 29Items-Projects/7Algorithms/7-String-algorithms/src/main/java/com/stringalgorithms/algorithms/LongestCommonSubstring.java
package com.stringalgorithms.algorithms;

import java.util.Objects;

public final class LongestCommonSubstring {
    private LongestCommonSubstring() {
    }

    public static String find(String first, String second) {
        Objects.requireNonNull(first, "first must not be null");
        Objects.requireNonNull(second, "second must not be null");

        int[][] lengths = new int[first.length() + 1][second.length() + 1];
        int bestLength = 0;
        int bestEnd = 0;

        for (int i = 1; i <= first.length(); i++) {
            for (int j = 1; j <= second.length(); j++) {
                if (first.charAt(i - 1) == second.charAt(j - 1)) {
                    lengths[i][j] = lengths[i - 1][j - 1] + 1;
                    if (lengths[i][j] > bestLength) {
                        bestLength = lengths[i][j];
                        bestEnd = i;
                    }
                }
            }
        }

        return first.substring(bestEnd - bestLength, bestEnd);
    }

    public static int length(String first, String second) {
        return find(first, second).length();
    }
}

File: 29Items-Projects/7Algorithms/7-String-algorithms/python/string_algorithms/algorithms/longest_common_substring.py
def longest_common_substring(first: str, second: str) -> str:
    lengths = [[0] * (len(second) + 1) for _ in range(len(first) + 1)]
    best_length = 0
    best_end = 0

    for i, first_char in enumerate(first, start=1):
        for j, second_char in enumerate(second, start=1):
            if first_char == second_char:
                lengths[i][j] = lengths[i - 1][j - 1] + 1
                if lengths[i][j] > best_length:
                    best_length = lengths[i][j]
                    best_end = i

    return first[best_end - best_length : best_end]


def longest_common_substring_length(first: str, second: str) -> int:
    return len(longest_common_substring(first, second))
